<a href="https://colab.research.google.com/github/Mamolinac/Business-intelligence/blob/main/1_Reto1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Se carga las librerías de trabajo


In [ ]:
import numpy as np #Posee funciones matemáticas
import pandas as pd #Comunicación con el excel



1. Se carga la base de datos

In [ ]:
nxl='/content/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl)
XDB.head(100)


ID  Age  Urban  Mother_Edu  Father_Edu  Siblings  Black  Hispanic  White  \
0     1   21    1.0         8.0         8.0       1.0    0.0       0.0    1.0   
1     2   20    1.0         5.0         8.0       8.0    0.0       0.0    1.0   
2     3   18    1.0        10.0        12.0       3.0    0.0       0.0    1.0   
3     4   17    1.0        11.0        12.0       3.0    0.0       0.0    1.0   
4     5   20    1.0        12.0        12.0       1.0    0.0       0.0    1.0   
..  ...  ...    ...         ...         ...       ...    ...       ...    ...   
95   96   18    1.0         6.0         NaN       2.0    0.0       0.0    1.0   
96   97   21    1.0        12.0        11.0       7.0    0.0       0.0    1.0   
97   98   22    1.0        17.0        14.0       2.0    0.0       0.0    1.0   
98   99   22    1.0        12.0        16.0       1.0    0.0       0.0    1.0   
99  100   16    1.0         NaN         NaN       2.0    0.0       0.0    1.0   

    Christian  ...  Height  Weight  Outgoing_Kid  Outgoing_Adult  HealthPlan  \
0         1.0  ...    65.0     NaN           NaN             NaN         NaN   
1         1.0  ...    62.0   120.0           0.0             1.0         1.0   
2         1.0  ...     NaN     NaN           1.0             1.0         1.0   
3         1.0  ...    67.0   110.0           0.0             1.0         NaN   
4         1.0  ...    63.0   130.0           NaN             NaN         NaN   
..        ...  ...     ...     ...           ...             ...         ...   
95        1.0  ...    64.0   122.0           0.0             1.0         1.0   
96        1.0  ...    66.0   134.0           1.0             1.0         1.0   
97        1.0  ...    63.0   125.0           1.0             1.0         1.0   
98        1.0  ...     NaN     NaN           0.0             1.0         1.0   
99        0.0  ...    64.0   120.0           0.0             1.0         1.0   

     Income  Marital_Status  Education  WeeksEmployed  NumberSpouses  
0       NaN             NaN        NaN            NaN            NaN  
1       0.0             1.0       12.0            0.0            1.0  
2       0.0             1.0       12.0           52.0            1.0  
3       NaN             NaN        NaN            NaN            NaN  
4       NaN             NaN        NaN            NaN            NaN  
..      ...             ...        ...            ...            ...  
95  38000.0             0.0       12.0           52.0            1.0  
96  28000.0             1.0       14.0           52.0            2.0  
97      0.0             1.0       13.0            0.0            1.0  
98   4000.0             1.0       14.0           52.0            1.0  
99  43000.0             3.0       12.0           52.0            1.0  

[100 rows x 24 columns]

Seleccionar las variables

In [ ]:
XDB=XDB[['Age','Urban','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]
XDB.head(10)

Age  Urban  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  \
0   21    1.0         8.0         8.0           5          NaN    65.0   
1   20    1.0         5.0         8.0           5         16.0    62.0   
2   18    1.0        10.0        12.0           5         20.0     NaN   
3   17    1.0        11.0        12.0           5          NaN    67.0   
4   20    1.0        12.0        12.0           4         23.0    63.0   
5   19    1.0        12.0        12.0           4         27.0    64.0   
6   15    1.0        12.0        12.0           3         26.0    65.0   
7   21    1.0         9.0         6.0           3         23.0    65.0   
8   16    1.0        12.0        10.0           6         26.0    66.0   
9   19    1.0        12.0        12.0           3         19.0    66.0   

   Weight   Income  
0     NaN      NaN  
1   120.0      0.0  
2     NaN      0.0  
3   110.0      NaN  
4   130.0      NaN  
5   200.0  40000.0  
6   131.0  25000.0  
7   179.0  27400.0  
8   145.0  52000.0  
9   115.0      NaN

In [ ]:
#Se cuenta el número de personas encuestadas
total_respondents = len(XDB)

print(f"Número total de personas que respondieron: {total_respondents}")

Número total de personas que respondieron: 12686


In [ ]:
# Porcentaje de cada categoría en Urban
categories = XDB['Urban'].value_counts(normalize=True) * 100
print('Porcentaje de datos en cada categoría:\n', categories)
#1.0 → 79.16%: Significa que el 79.16% de las personas en la base de datos viven en áreas urbanas.
#0.0 → 20.83%: Indica que el 20.83% de las personas viven en áreas rurales.

Porcentaje de datos en cada categoría:
 Urban
1.0    78.839159
0.0    21.160841
Name: proportion, dtype: float64


In [ ]:
# Eliminar filas con al menos un valor faltante
XDB = XDB.dropna()
XDB.head(10)




Age  Urban  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  \
1    20    1.0         5.0         8.0           5         16.0    62.0   
5    19    1.0        12.0        12.0           4         27.0    64.0   
6    15    1.0        12.0        12.0           3         26.0    65.0   
7    21    1.0         9.0         6.0           3         23.0    65.0   
8    16    1.0        12.0        10.0           6         26.0    66.0   
12   21    1.0        12.0        16.0           5         25.0    71.0   
13   16    1.0        12.0        12.0           5         21.0    67.0   
14   15    1.0        12.0        12.0           4         23.0    73.0   
15   21    1.0        12.0        12.0           4         25.0    63.0   
16   22    1.0        12.0        15.0           2         24.0    69.0   

    Weight   Income  
1    120.0      0.0  
5    200.0  40000.0  
6    131.0  25000.0  
7    179.0  27400.0  
8    145.0  52000.0  
12   180.0  60000.0  
13   135.0  48000.0  
14   185.0      0.0  
15   130.0  38000.0  
16   160.0  48000.0

In [ ]:
total_respondents = len(XDB)
print(f"Número total de personas que respondieron: {total_respondents}")

Número total de personas que respondieron: 5948


In [ ]:
# Porcentaje de cada categoría en "Urban"
categories = XDB['Urban'].value_counts(normalize=True) * 100
print('Porcentaje de datos en cada categoría:\n', categories)
#1.0 → 79.16%: Significa que el 79.16% de las personas en la base de datos viven en áreas urbanas.
#0.0 → 20.83%: Indica que el 20.83% de las personas viven en áreas rurales.

Porcentaje de datos en cada categoría:
 Urban
1.0    79.169469
0.0    20.830531
Name: proportion, dtype: float64


Se separan las variables de entrada y salida

In [ ]:
#Separan las variables de entrada y salida
XD=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de entrada
YD=XDB[['Urban']] #Variable de salida-Variable Pronóstico

Se comienza la implementación del modelo

In [ ]:
# Intervalo de la Edad
age_min = XDB['Age'].min()
age_max = XDB['Age'].max()
age_interval = (age_min, age_max)

# Intervalo de la Altura
height_min = XDB['Height'].min()
height_max = XDB['Height'].max()
height_interval = (height_min, height_max)

# Intervalo del Peso
weight_min = XDB['Weight'].min()
weight_max = XDB['Weight'].max()
weight_interval = (weight_min, weight_max)

# Imprimir los resultados
print(f"Intervalo de la Edad: {age_interval}")
print(f"Intervalo de la Altura: {height_interval}")
print(f"Intervalo del Peso: {weight_interval}")


# Filtrar los datos para el área urbana (Urban = 1)
datos_urbanos = XDB[XDB['Urban'] == 1]

# Calcular la media y la desviación estándar de la altura en el área urbana
media_altura_urbana = datos_urbanos['Height'].mean()
desviacion_estandar_altura_urbana = datos_urbanos['Height'].std()

# Calcular los límites superior e inferior
limite_superior = media_altura_urbana + desviacion_estandar_altura_urbana
limite_inferior = media_altura_urbana - desviacion_estandar_altura_urbana

# Imprimir los resultados
print("Límite superior de la altura en el área urbana:", limite_superior)
print("Límite inferior de la altura en el área urbana:", limite_inferior)

# Filtrar los datos para el área rural (Urban = 0)
datos_rurales = XDB[XDB['Urban'] == 0]

# Calcular la media y la desviación estándar del peso en el área rural
media_peso_rural = datos_rurales['Weight'].mean()
desviacion_estandar_peso_rural = datos_rurales['Weight'].std()

# Calcular los límites superior e inferior
limite_superior = media_peso_rural + desviacion_estandar_peso_rural
limite_inferior = media_peso_rural - desviacion_estandar_peso_rural

# Imprimir los resultados
print("Límite superior del peso en el área rural:", limite_superior)
print("Límite inferior del peso en el área rural:", limite_inferior)

Intervalo de la Edad: (15, 22)
Intervalo de la Altura: (51.0, 80.0)
Intervalo del Peso: (72.0, 375.0)
Límite superior de la altura en el área urbana: 70.96360588384263
Límite inferior de la altura en el área urbana: 62.85971116861012
Límite superior del peso en el área rural: 179.34365004593002
Límite inferior del peso en el área rural: 118.66280677408612


In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB() #Aloja el modelo
mnb.fit(XD,YD) #Entrena el modelo

#Obtenemos la información del modelo
u=mnb.theta_ #Obtenemos la media de las variables
print("La media de las variables es:\n",u)
var=mnb.var_ #Se calculan las varianzas
sigma=np.sqrt(var) #Se calcula la desviación estándar
print("La desviación estándar de los ingresos es:\n",sigma)

#Se determinan los límites superiores e inferiores de los datos
LS=u+sigma
LI=u-sigma
print("El límite superior de las variables es:\n",LS)
print("El límite inferior de las variables es:\n",LI)

#Evaluamos un solicitante
XDS=[[34,12,15,2,24,69,160,48000]] #Como un valor de entrada-Ingresos
ydp=mnb.predict(XDS) #Se obtiene la prediccion
print("La predicción del encuestado es:\n", ydp)

if ydp==0:
  print("El solicitante vive en zona rural")
if ydp==1:
  print("El solicitante vive en zona urbana")

La media de las variables es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La desviación estándar de los ingresos es:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
El límite superior de las variables es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.12309482e+00
  2.67720628e+01 7.11135895e+01 1.74779433e+02 6.93634044e+04]]
El límite inferior de las variables es:
 [[ 1.58913978e+01  7.57462359e+00  6.38281079e+00  2.18703841e+00
   1.77786493e+01  6.316

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Evaluamos el desempeño del modelo

In [ ]:
from re import S
from sklearn.metrics import confusion_matrix

#Haga un barrido de los datos
ydp2=mnb.predict(XD) #Haga el pronostico para los datos base
cm=confusion_matrix(YD,ydp2) #Realiza la matriz de confusión
print("La matriz de confusión es:\n",cm)
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

#Métricas desempeño
#1.Exactitud
Ex=(VP+VN)/(VP+VN+FP+FN) #Exactitud: comportamiento general urbano
print("La exactitud del modelo es:\n",Ex)

TE=(FN+FP)/(VP+VN+FP+FN) #Tasa de error: comportamiento general rural
print("La tasa de error del modelo es:\n",TE)

#2 Sensibilidad

Se=VP/(VP+FN) #Sensibilidad: comportamiento correcto urbano
print("La sensibilidad del modelo es:\n",Se)

#3 Especifidad

Sp=VN/(VN+FP) #Especificidad: comportamiento correcto rural
print("La especificidad del modelo es:\n",Sp)

#4 La precisión
Pr=VP/(VP+FP) #Precisión: comportamiento correcto urbano #Sobre el total de los precedidos

print("La precisión del modelo es:\n",Pr)

#5 La precisión negativa
PrNeg=VN/(VN+FN) #Precisión negativa: comportamiento correcto rural #Sobre el total de los no precedidos

print("La precisión negativa del modelo es:\n",PrNeg)





La matriz de confusión es:
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es:
 0.7908540685944856
La tasa de error del modelo es:
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367
La precisión negativa del modelo es:
 0.2727272727272727


Con base en la información suministrada, se desarrolló un modelo para determinar si una persona reside en una zona urbana o rural. Para ello, se consideraron variables socioeconómicas como edad, educación de los padres, estatura, peso, ingresos, escala de autoestima y tamaño de la familia.

La base de datos original contenía un total de 12.686 registros. Sin embargo, tras un proceso de depuración para eliminar las respuestas incompletas, se obtuvo un conjunto final de 5.948 datos, de los cuales el 79,16% corresponde a personas que viven en zonas urbanas y el 20,83% en zonas rurales.


El modelo alcanzó una **precisión** promedio del 79,08%, lo que indica un desempeño aceptable en la clasificación de la zona de residencia para la zona urbana. Sin embargo, la **especificidad** fue del 0,24%, lo que refleja una mayor dificultad en la identificación de residencia rurales. A pesar de esto, el **exactitud** del modelo se mantuvo cercano al 79%, evidenciando un criterio sólido en la determinación residencia para la zona urbana. Lo que nos da entender que el modelo es excelente para la clasificación de la zona de residencia para la zona urbana.

